go-controller/pkg/ovn/policy.go

```diff
@@ -1148,276 +1159,298 @@ func hasAnyLabelSelector(peers []knet.NetworkPolicyPeer) bool {
 			policy.Namespace, policy.Name)
-		return nil
 	}
-	np := NewNetworkPolicy(policy)
+	aclLogging := nsInfo.aclLogging
+	nsUnlock()
 
-	if len(nsInfo.networkPolicies) == 0 {
-		err := oc.addPolicyToDefaultPortGroups(np, &nsInfo.aclLogging)
+	// 2. create network policy without ns Lock, cleanup on failure
+	npKey := getPolicyKey(policy)
+	var np *networkPolicy
+	var err error
+
+	np, err = oc.createNetworkPolicy(policy, &aclLogging)
+	defer func() {
 		if err != nil {
-			nsUnlock()
-			return fmt.Errorf("adding network policy %s/%s failed: %v",
-				policy.Namespace, policy.Name, err)
-		}
-		defer func() {
-			if err != nil {
-				errDelete := oc.delPolicyFromDefaultPortGroups(np)
-				// rollback failed, best effort cleanup; won't add to retry mechanism since item doesn't exist in cache yet.
-				if errDelete != nil {
-					klog.Warningf("Rollback of default port groups and acls for policy: %s/%s failed, Unable to ensure namespace for network policy: error %v", policy.Namespace, policy.Name, errDelete)
-					return
+			// try to cleanup network policy straight away
+			// it will be retried later with add/delete network policy handlers if it fails
+			cleanupErr := oc.networkPolicies.DoWithLock(npKey, func(npKey string) error {
+				np, ok := oc.networkPolicies.Load(npKey)
+				if !ok {
+					klog.Infof("Deleting policy that is already deleted")
+					return nil
 				}
+				return oc.cleanupNetworkPolicy(np)
+			})
+			if cleanupErr != nil {
+				klog.Infof("Cleanup for failed create network policy returned an error: %v", cleanupErr)
 			}
-		}()
-	}
-	aclLogging := nsInfo.aclLogging
-	nsUnlock()
-	if err := oc.createNetworkPolicy(np, policy, &aclLogging); err != nil {
-		return fmt.Errorf("failed to create Network Policy: %s/%s, error: %v",
+		}
+	}()
+	if err != nil {
+		return fmt.Errorf("failed to create Network Policy %s/%s: %v",
 			policy.Namespace, policy.Name, err)
 	}
 
-	// Now do nsinfo operations to set the policy
-	nsInfo, nsUnlock, err = oc.ensureNamespaceLocked(policy.Namespace, false, nil)
-	if err != nil {
-		// rollback network policy
-		if err := oc.deleteNetworkPolicy(policy, np); err != nil {
-			// rollback failed, add to retry to cleanup
-			key := getPolicyKey(policy)
-			oc.retryNetworkPolicies.DoWithLock(key, func(key string) {
-				oc.retryNetworkPolicies.initRetryObjWithDelete(policy, key, np, false)
-			})
-		}
-		return fmt.Errorf("unable to ensure namespace for network policy: %s, namespace: %s, error: %v",
-			policy.Name, policy.Namespace, err)
+	// 3. lock namespace
+	nsInfo, nsUnlock = oc.getNamespaceLocked(policy.Namespace, false)
+	if nsInfo == nil {
+		// namespace was deleted while we were adding network policy,
+		// try to cleanup network policy
+		// expect retry to be handled by delete event that should come
+		klog.Warningf("Unable to get namespace %s at the end of network policy %s creation: %v",
+			policy.Namespace, policy.Name, err)
+		return err
```

Comment:
this is always nil, dont you want to return an error so then it will try to clean the network policy in the defer?

Reply:
of course I want, thanks for catching this :)